ucp/_libs/ucx_api.pyx

```diff
@@ -254,39 +311,51 @@ cdef class UCXWorker:
         return ucx_ep_create(ucp_ep, self)
 
 
-cdef class UCXEndpoint:
+def _ucx_endpoint_finalizer(handle_as_int, worker, inflight_msgs):
+    cdef ucp_ep_h handle = <ucp_ep_h><uintptr_t> handle_as_int
+    cdef ucs_status_ptr_t status
+
+    # Cancel all inflight messages
+    for msg in list(inflight_msgs.values()):
+        logger.debug("Future cancelling: %s" % msg["log_msg"])
+        # Notice, `request_cancel()` evoke the send/recv callback functions
+        worker.request_cancel(msg["ucp_request"])
+
+    # Close the endpoint
+    # TODO: Support UCP_EP_CLOSE_MODE_FORCE
+    status = ucp_ep_close_nb(handle, UCP_EP_CLOSE_MODE_FLUSH)
+    if UCS_PTR_IS_PTR(status):
+        ucp_request_free(status)
+    elif UCS_PTR_STATUS(status) != UCS_OK:
+        msg = ucs_status_string(UCS_PTR_STATUS(status)).decode("utf-8")
+        raise UCXError("Error while closing endpoint: %s" % msg)
+
+
+cdef class UCXEndpoint(UCXObject):
     """Python representation of `ucp_ep_h`
     Please use `ucx_ep_create()` to contruct an instance of this class
     """
     cdef:
-        object __weakref__
         ucp_ep_h _handle
+        dict _inflight_msgs
 
     cdef readonly:
-        bint initialized
-
-    cdef public:
         UCXWorker worker
 
-    def __cinit__(self, worker):
-        self._handle = NULL
-        self.worker = worker
-        self.initialized = False
+    cdef _init(self, UCXWorker worker, ucp_ep_h handle):
```

Comment:
Why not use `__cinit__` or perhaps `__init__` for this?

Reply:
Cython doesn't support a `cdef` constructor so it complains that `ucp_ep_h handle` isn't convertible to a `PyObject`.

Reply:
One way to workaround this limitation is take a `uintptr_t` instead (Python will allow this and perform type checking for it) and then cast to your specific pointer type.

Reply:
Proposed some changes in a more recent review to address this.